_posts/2023-06-13-http4s-authentication-part2.md

```diff
@@ -0,0 +1,883 @@
+
+    def send2FA(sendto: User): IO[Either[String,Response]] = {
+        val from = new Email("hkateu@gmail.com")
+        val to = new Email(sendto.email)
+
+        val subject = "Sending with SendGrid is Fun"
+        val message = s"""
+            |<H3>Website Login</H3>
+            |<p>Please complete the login process using google authenticator</p>
+            |<p>Scan the QR Barcode attached and send us the token number.</p>
+            """.stripMargin
+
+        val file = Paths.get("barCode.png")
+        val content = new Content("text/html", message)
+
+        val mail = new Mail(from, subject, to, content)
+        val attachments = new Attachments()
+        attachments.setFilename(file.getFileName().toString())
+        attachments.setType("application/png")
+        attachments.setDisposition("attachment")
+        val attachmentContentBytes = Files.readAllBytes(file)
+        val attachmentContent = Base64.getEncoder().encodeToString(attachmentContentBytes)
+        attachments.setContent(attachmentContent)
+        mail.addAttachments(attachments)
+
+        val sg = new SendGrid(Properties.envOrElse("SENDGRID_API_KEY","undefined"))
+        val request = new Request()
+        val response: IO[Either[String,Response]] =
+            IO{
+                Try {
+                    request.setMethod(Method.POST)
+                    request.setEndpoint("mail/send")
+                    request.setBody(mail.build())
+                    val res = sg.api(request)
+                    res
+                }.toEither.left.map(_.getMessage)
+            }
+
+        response
+    }
+}
+```
+
+**Http4s Routes and Server.**
+
+In this section, we create the routes and server using http4s. Let's create a new scala file called OtpAuth.scala and add the following.
+
+```scala
+import cats.effect.{IOApp, IO, ExitCode}
+import com.rockthejvm.OtpInteractiveService
+import com.rockthejvm.User
+import com.rockthejvm.Generator.*
+
+object OtpAuth extends IOApp:
+    val user = User("john", "john@email.com")
+    val tokenService = new OtpInteractiveService(Hotp, user)
+
+    override def run(args: List[String]): IO[ExitCode] =  ???
+```
+
+Here we create an instance of `OtpInteractiveService` to which we supply a `Generator` of type `Hotp` and a `User` as constructor arguments.
+
+```scala
+import org.http4s.*
+import org.http4s.dsl.io.*
+import scala.util.*
+    ...
+    val routes: HttpRoutes[IO] =
+        HttpRoutes.of[IO]{
+            case GET -> Root / "login" =>
+                tokenService.otpService.makeBarCodeImg.flatMap{value =>
+                    value match
+                        case Right(_) =>
+                            tokenService.otpService.getToken().map(println) >>
+                            tokenService.send2FA(user).flatMap{response =>
+                                response match
+                                    case Right(res) =>
+                                        IO(println(res.getBody())) >>
```

Comment:
Would be great to use a logger instead of prints - can use log4cats for it

Reply:
Fixed